### PR TITLE
tests: fix type tests for Link, make sure type tests fail in ci and add type tests for union `to` and difference in params

### PR DESCRIPTION
**Pull request by chorobin**
- Tests were failing and were not type checking in ci
- Added test cases for `to` being a union, params being optional if `from` and `to` contain the same params

### CODE REVIEW

File: packages/react-router/tests/link.test-d.tsx [hunk 1]
```diff
@@ -216,6 +230,116 @@ test('when navigating from a route with no params to a route with params, params
...
+  const params = expectTypeOf(TestLink)
+    .parameter(0)
+    .toHaveProperty('params')
+    .extract<Function>()
+
+  params.returns.branded.toEqualTypeOf<{ invoiceId?: string | undefined }>()
+  params.parameter(0).toEqualTypeOf<{ invoiceId: string }>()
+})
+
+test('when navigating to a union of routes, params is optional', () => {
```
Comment (schiller-manuel) [inline packages/react-router/tests/link.test-d.tsx:253]: is this correct though? shouldn't `params` be `{ invoiceId: string } | `{ postId: string}`?

File: packages/react-router/tests/link.test-d.tsx [hunk 2]
```diff
@@ -216,6 +230,116 @@ test('when navigating from a route with no params to a route with params, params
...
+  const TestLink = Link<
+    RouteTree,
+    string,
+    '/invoices/$invoiceId/' | '/posts/$postId/'
+  >
+  expectTypeOf(TestLink).parameter(0).not.toMatchTypeOf<{ params: unknown }>()
```
Comment (schiller-manuel) [inline packages/react-router/tests/link.test-d.tsx:259]: is this the type check for "params is optional"? how? and is there a more explicit way?

File: packages/react-router/tests/link.test-d.tsx (hunk 1 above)
Comment (schiller-manuel) [reply packages/react-router/tests/link.test-d.tsx:253]: yeah I was expecting this test:

https://github.com/TanStack/router/pull/1418/files#diff-921603b68301ae5e52534582aeb75f7defc2946a96ad3b268b28ad34ae5fec04R262

why would we need this test here? the test below already checks everything, right?

File: packages/react-router/tests/link.test-d.tsx [hunk 3]
```diff
@@ -216,6 +230,116 @@ test('when navigating from a route with no params to a route with params, params
...
+  params.returns.branded.toEqualTypeOf<
+    { invoiceId: string } | { postId: string }
+  >()
+
+  params
+    .parameter(0)
+    .toEqualTypeOf<{} | { invoiceId: string } | { postId: string }>()
+})
+
+test('when navigating to a union of routes including the route, params is optional', () => {
```
Comment (schiller-manuel) [inline packages/react-router/tests/link.test-d.tsx:297]: "including the route" .. which one?

Comment (schiller-manuel) [reply packages/react-router/tests/link.test-d.tsx:297]: ah you meant "the root", right?

File: packages/react-router/tests/link.test-d.tsx [hunk 4]
```diff
@@ -304,3 +428,83 @@ test('when navigating to a route with search params, search params is a function
...
+    .parameter(0)
+    .toHaveProperty('search')
+    .extract<Function>()
+
+  params.returns.branded.toEqualTypeOf<{ page: number } | {}>()
+
+  params.parameter(0).toEqualTypeOf<{} | { page: number }>()
+})
+
+test('when navigating to a union of routes including the route, search params is optional', () => {
```
Comment (schiller-manuel) [inline packages/react-router/tests/link.test-d.tsx:472]: "the route" => "the root"

File: packages/react-router/tests/link.test-d.tsx (hunk 1 above)
Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:253]: I will double check but I believe this is slightly different as we're checking if it's optional. But it's a good point if this is intentional behaviour or not. Should it be optional for a union of different routes?

I find the unions a harder use case to think about if it should be optional or not. For example the default case `string` falls under the same code path and this makes sense to be optional

Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:253]: `string` is basically considered to be any route and is therefore a union of everything

File: packages/react-router/tests/link.test-d.tsx (hunk 2 above)
Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:259]: Docs cover this

https://github.com/mmkal/expect-type#error-messages

File: packages/react-router/tests/link.test-d.tsx (hunk 3 above)
Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:297]: Yep. Sorry, bad spelling

File: packages/react-router/tests/link.test-d.tsx (hunk 2 above)
Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:259]: I'm going to fix this comments tomorrow so don't merge. But you have me doubting myself so I will check

File: packages/react-router/tests/link.test-d.tsx (hunk 4 above)
Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:472]: Yep. Same mistake!

**Review by SeanCassiere — COMMENTED** [review]
Correct me if I'm wrong, but I believe we aren't testing for external links.

The implementation for the `<Link>` component checks for external links and sets the `href` property to whatever the input of `to` was.

Not sure if that's something that'll be covered by this PR or not, just putting it out there since the types for Link are being touched here.

**Comment by chorobin** [comment]
> Correct me if I'm wrong, but I believe we aren't testing for external links.
> 
> The implementation for the `<Link>` component checks for external links and sets the `href` property to whatever the input of `to` was.
> 
> Not sure if that's something that'll be covered by this PR or not, just putting it out there since the types for Link are being touched here.

Maybe we can do sanity check that we allow any string literal type in `to`. But bare in mind these are only type tests and doesn't test the runtime behaviour. We need both but since I'm touching the types most of the time, that's what I see break the most so I wanted to add some

File: packages/react-router/tests/link.test-d.tsx (hunk 2 above)
Comment (chorobin) [reply packages/react-router/tests/link.test-d.tsx:259]: I'm trying to think of a different way but it's difficult as the types can be quite complex for `params`

**Pull request merged**